### PR TITLE
kubeadm: remove ClusterConfiguration.DNS.Type from v1beta3

### LINKED ISSUE #2398: remove ClusterConfiguration.DNS.Type
https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#DNS

with the removal of kube-dns:
https://github.com/kubernetes/kubernetes/pull/99646

kubeadm no longer supports multiple DNS types:
this is scheduled for v1beta3.

### PR DESCRIPTION
**Pull request by neolit123**
#### What this PR does / why we need it:

Kubeadm no longer supports kube-dns and CoreDNS is the only
supported DNS server. Remove ClusterConfiguration.DNS.Type
from v1beta3 that is used to set the DNS server type.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/2398
xref https://github.com/kubernetes/kubeadm/issues/1796

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: remove the field ClusterConfiguration.DNS.Type in v1beta3 since CoreDNS is the only supported DNS type.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
/hold
for review

**Review by pacoxu — APPROVED** [review]
/lgtm

**Comment by pacoxu** [comment]
As node-local-dns is GAed since 1.18, is there a plan to add node-local-dns to ClusterConfiguration.DNS.Cache=true in kubeadm?

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:52]: i guess this can just be `kubeadm.CoreDNS` instead of using this literal string.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/types.go:165]: Should we remove the field from internal types as well?
AFAIK the code for generating kube-dns manifest has been already removed, so we don't really need this anymore. Am I wrong?

**Comment by neolit123** [comment]
@pacoxu 

> As node-local-dns is GAed since 1.18, is there a plan to add node-local-dns to ClusterConfiguration.DNS.Cache=true in kubeadm?

we had a few discussions about this in the past.
https://github.com/kubernetes/kubeadm/issues/1220

as an addon, my main argument against it is that its not something that is exactly required for a cluster to work.
kubeadm tries to be minimal and only apply essentials and security defaults. the node local cache would be something that is like a "nice to have", but not essential.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:165]: i'd happily remove it, but i don't think it would be possible to round trip a v1beta2 -> internal -> v1beta2 without this field.

Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go:165]: if the problem are roundtrips you can force the fuzzer to set DNS.Type to empty string, so the roundtrip  should pass

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:165]: discussed the problem with @fabriziopandini offline.
created a related PR here: https://github.com/kubernetes/kubernetes/pull/101598

i will update this one too, soon.

**Comment by neolit123** [comment]
@fabriziopandini i have updated the PR. looking for LGTM.

getting this right is difficult and other fuzzer magic did not help.

the scenario around DNS.Type unravels some limitations in the fuzzers + stock API machinery:
- field exists in the old public API
- field is removed in the latest public API
- field must exists in the internal type, since validation is performed on internal types only and until the old public API is removed.

i argue that this model is flawed and it forces API maintainers to have their latest public API converters and internal APIs know about outdated fields (see cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go).

**Comment by fabriziopandini** [comment]
/lgtm

We can eventually iterate on the fuzzer/internal types on separated PRs given that this isn't user facing.
IMO, what is really important in the long term, is that we find a way to untangle defaulting & the fuzzing logic

**Pull request merged**
